# SQL Project

# Description

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers.
You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

Plan:
    
Describe the goals of the study.

Study the tables (print the first rows).

Make an SQL query for each of the tasks.

Output the results of each query in the Notebook.

Describe your conclusions for each of the tasks.

Task:
    
Find the number of books released after January 1, 2000.

Find the number of user reviews and the average rating for each book.

Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

Identify the author with the highest average book rating: look only at books with at least 50 ratings.

Find the average number of text reviews among users who rated more than 50 books.

## Downloading data

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
def func_display(x):
    x = pd.io.sql.read_sql(x, con = engine)
    display(x.head())

In [3]:
books = '''
SELECT * FROM
    books
'''
func_display(books)

authors = '''
SELECT * FROM
    authors
'''
func_display(authors)

publishers = '''
SELECT * FROM
    publishers
'''
func_display(publishers)

ratings = '''
SELECT * FROM
    ratings
'''
func_display(ratings)

reviews = '''
SELECT * FROM
    reviews
'''
func_display(reviews)



book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [4]:
query = '''
SELECT *
FROM books
'''
books_count = pd.read_sql(query, engine)
print(books_count.duplicated().sum())

0


In [5]:
query_1 = '''
SELECT *
FROM authors
'''
authors_count = pd.read_sql(query_1, engine)
print(authors_count.duplicated().sum())

0


In [6]:
query_2 = '''
SELECT *
FROM publishers
'''
publishers_count = pd.read_sql(query_2, engine)
print(publishers_count.duplicated().sum())

0


In [7]:
query_3 = '''
SELECT *
FROM ratings
'''
ratings_count = pd.read_sql(query_3, engine)
print(ratings_count.duplicated().sum())

0


In [8]:
query_4 = '''
SELECT *
FROM reviews
'''
reviews_count = pd.read_sql(query_4, engine)
print(reviews_count.duplicated().sum())

0


## Find the number of books released after January 1, 2000.

In [9]:
query1 = '''
SELECT COUNT (book_id)
FROM books
'''
books_count = pd.read_sql(query1, engine)
print(books_count)

   count
0   1000


Total number of books 1000

In [10]:
query2 = '''
SELECT COUNT (book_id)
FROM books
WHERE publication_date > '2000-01-01'
'''
books_count2000 = pd.read_sql(query2, engine)
print(books_count2000)

   count
0    819


Number of books released after 2000/1/1 - 819 books

## Find the number of user reviews and the average rating for each book.

In [11]:
query_3 = '''
SELECT COUNT(DISTINCT review_id) AS reviews_count, 
                    AVG(rating) AS avg_rating, 
                    books.book_id
FROM(( books
LEFT JOIN ratings ON ratings.book_id = books.book_id)
LEFT JOIN reviews ON reviews.book_id = ratings.book_id)
GROUP BY books.book_id
'''
avg_rating = pd.read_sql(query_3, engine)
print(avg_rating)

     reviews_count  avg_rating  book_id
0                2    3.666667        1
1                1    2.500000        2
2                3    4.666667        3
3                2    4.500000        4
4                4    4.000000        5
..             ...         ...      ...
995              3    3.666667      996
996              3    3.400000      997
997              4    3.200000      998
998              2    4.500000      999
999              4    3.833333     1000

[1000 rows x 3 columns]


## Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

In [12]:
query5 = '''
SELECT publisher, COUNT(books.book_id)
FROM 
    publishers
INNER JOIN books ON books.publisher_id = publishers.publisher_id
WHERE num_pages > 50
GROUP BY publishers.publisher
ORDER BY COUNT(book_id) DESC
'''
publisher_name = pd.read_sql(query5, engine)
display(publisher_name.head(1))

publisher  count
0  Penguin Books     42

Top publisher is Penguin Books.

## Identify the author with the highest average book rating: look only at books with at least 50 ratings.

In [13]:
query6 = '''
SELECT a.author, AVG(r.rating) AS average_rating
FROM books b
JOIN ratings r ON b.book_id = r.book_id
JOIN authors a ON b.author_id = a.author_id
WHERE b.book_id IN (
  SELECT book_id
  FROM ratings
  GROUP BY book_id
  HAVING COUNT(rating_id) > 50
)
GROUP BY a.author_id, a.author
ORDER BY average_rating DESC;
'''
top_author = pd.read_sql(query6, engine)
display(top_author.head())

author  average_rating
0         J.K. Rowling/Mary GrandPré        4.287097
1  Markus Zusak/Cao Xuân Việt Khương        4.264151
2                     J.R.R. Tolkien        4.246914
3                  Louisa May Alcott        4.192308
4                       Rick Riordan        4.080645

J.K. Rowling/Mary GrandPré	has the highest average book rating.

## Find the average number of text reviews among users who rated more than 50 books.

In [22]:
query7 = '''
SELECT AVG(num_reviews) AS avg_num_reviews
FROM (
  SELECT r.username, COUNT(*) AS num_reviews
  FROM reviews r
  JOIN (
    SELECT username
    FROM ratings
    GROUP BY username
    HAVING COUNT(DISTINCT book_id) > 50
  ) u ON r.username = u.username
  GROUP BY r.username
)subquery;
'''
num_reviews = pd.read_sql(query7, engine)
display(num_reviews.head())

avg_num_reviews
0        24.333333